src/mmds/src/lib.rs

```diff
@@ -78,108 +87,80 @@ fn sanitize_uri(mut uri: String) -> String {
-                        )
-                    }
-                    MmdsError::UnsupportedValueType => build_response(
-                        request.http_version(),
-                        StatusCode::NotImplemented,
-                        Body::new(e.to_string()),
-                    ),
-                    MmdsError::NotInitialized => unreachable!(),
-                },
+                    StatusCode::NotFound,
+                    Body::new(error_msg),
+                )
             }
-        }
-        Err(e) => match e {
-            RequestError::InvalidHttpVersion(err_msg) => build_response(
-                Version::default(),
-                StatusCode::NotImplemented,
-                Body::new(err_msg.to_string()),
-            ),
-            RequestError::InvalidUri(err_msg) => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new(err_msg.to_string()),
-            ),
-            RequestError::InvalidHttpMethod(err_msg) => build_response(
-                Version::default(),
+            MmdsError::UnsupportedValueType => build_response(
+                request.http_version(),
                 StatusCode::NotImplemented,
-                Body::new(err_msg.to_string()),
-            ),
-            RequestError::InvalidRequest => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new("Invalid request.".to_string()),
+                Body::new(e.to_string()),
             ),
-            RequestError::InvalidHeader => build_response(
-                Version::default(),
-                StatusCode::BadRequest,
-                Body::new("Invalid headers.".to_string()),
-            ),
-            // `micro-http` supports a predefined list of HTTP headers.
-            // It shouldn't reach this point, because it ignores the
-            // HTTP unsupported headers.
-            RequestError::UnsupportedHeader => unreachable!(),
+            MmdsError::NotInitialized => unreachable!(),
         },
     }
 }
 
 #[cfg(test)]
 mod tests {
     extern crate serde_json;
+
     use super::*;
+    use dumbo::tcp::parse_request_bytes;
+    use micro_http::{Request, Response, StatusCode, Version};
 
     fn check_http_method_failure(method: String, status: StatusCode, err_msg: String) {
-        let request = format!("{} http://169.254.169.254/ HTTP/1.1\r\n\r\n", method);
+        let request_string = format!("{} http://169.254.169.254/ HTTP/1.1\r\n\r\n", method);
+        let request_bytes = request_string.as_bytes();
         let mut expected_response = Response::new(Version::Http11, status);
         expected_response.set_body(Body::new(err_msg));
-        let actual_response = parse_request(request.as_bytes());
+        if status == StatusCode::MethodNotAllowed {
+            expected_response.allow_method(Method::Get);
+        }
+        expected_response.allow_method(Method::Get);
```

Comment:
These lines conflicts themselves. You add an `Allow` header even if the expected response should not contain it (e.g. check method failure for "GET", StatusCode::Ok, "dont care"). Furthermore, the added Allow header is not asserted against the actual response.

```suggestion
        if status == StatusCode::MethodNotAllowed {
            expected_response.allow_method(Method::Get);
        }
```